File: go/Assembler/template_common/logger.go
package template_common

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// LogLevel represents the severity of a log message
type LogLevel int

const (
	DEBUG LogLevel = iota
	INFO
	WARN
	ERROR
	NONE
)

// LogRotation represents log rotation intervals
type LogRotation int

const (
	ROTATION_NONE LogRotation = iota
	ROTATION_HOURLY
	ROTATION_DAILY
)

var logLevelNames = map[LogLevel]string{
	DEBUG: "DEBUG",
	INFO:  "INFO ",
	WARN:  "WARN ",
	ERROR: "ERROR",
	NONE:  "NONE ",
}

var logLevelColors = map[LogLevel]string{
	DEBUG: "\033[90m",  // Gray
	INFO:  "\033[97m",  // White
	WARN:  "\033[93m",  // Yellow
	ERROR: "\033[91m",  // Red
}

const colorReset = "\033[0m"

// Logger is a simple custom logger for consistent logging across all language implementations
type LoggerType struct {
	currentLogLevel   LogLevel
	logFilePath       string
	consoleOutput     bool
	mutex             sync.Mutex
	logRotation       LogRotation
	baseLogFilePath   string
	currentRotatedPath string
	contextLogFiles   map[string]string
}

var globalLogger = &LoggerType{
	currentLogLevel: INFO,
	consoleOutput:   true,
	contextLogFiles: make(map[string]string),
}

// Configure sets up the logger with the specified level, file path, and console output
func Configure(level LogLevel, logFilePath string, consoleOutput bool, rotation LogRotation) {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()

	globalLogger.currentLogLevel = level
	globalLogger.baseLogFilePath = logFilePath
	globalLogger.consoleOutput = consoleOutput
	globalLogger.logRotation = rotation

	// Generate the initial rotated path
	globalLogger.currentRotatedPath = getRotatedFilePath()
	globalLogger.logFilePath = globalLogger.currentRotatedPath

	// Create or clear log file if specified
	if globalLogger.logFilePath != "" {
		timestamp := time.Now().Format("2006-01-02 15:04:05")
		header := fmt.Sprintf("=== Log started at %s ===\n", timestamp)
		err := os.WriteFile(globalLogger.logFilePath, []byte(header), 0644)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Failed to initialize log file: %v\n", err)
		}
	}
}

// getRotatedFilePath generates the rotated file path based on rotation setting
func getRotatedFilePath() string {
	if globalLogger.baseLogFilePath == "" {
		return ""
	}

	if globalLogger.logRotation == ROTATION_NONE {
		return globalLogger.baseLogFilePath
	}

	dir := filepath.Dir(globalLogger.baseLogFilePath)
	ext := filepath.Ext(globalLogger.baseLogFilePath)
	nameWithoutExt := strings.TrimSuffix(filepath.Base(globalLogger.baseLogFilePath), ext)

	now := time.Now()
	var suffix string

	if globalLogger.logRotation == ROTATION_HOURLY {
		suffix = now.Format("2006-01-02_15")
	} else { // ROTATION_DAILY
		suffix = now.Format("2006-01-02")
	}

	rotatedFileName := fmt.Sprintf("%s_%s%s", nameWithoutExt, suffix, ext)
	return filepath.Join(dir, rotatedFileName)
}

// SetLogLevel sets the current log level
func SetLogLevel(level LogLevel) {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()
	globalLogger.currentLogLevel = level
}

// GetLogLevel returns the current log level
func GetLogLevel() LogLevel {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()
	return globalLogger.currentLogLevel
}

// Debug logs a DEBUG level message
func Debug(message string, context ...string) {
	contextStr := ""
	if len(context) > 0 {
		contextStr = context[0]
	}
	log(DEBUG, message, contextStr)
}

// Info logs an INFO level message
func Info(message string, context ...string) {
	contextStr := ""
	if len(context) > 0 {
		contextStr = context[0]
	}
	log(INFO, message, contextStr)
}

// Warn logs a WARN level message
func Warn(message string, context ...string) {
	contextStr := ""
	if len(context) > 0 {
		contextStr = context[0]
	}
	log(WARN, message, contextStr)
}

// Error logs an ERROR level message
func Error(message string, context ...string) {
	contextStr := ""
	if len(context) > 0 {
		contextStr = context[0]
	}
	log(ERROR, message, contextStr)
}

// log is the core logging method
func log(level LogLevel, message string, context string) {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()

	if level < globalLogger.currentLogLevel {
		return
	}

	timestamp := time.Now().Format("2006-01-02 15:04:05.000")
	levelStr := logLevelNames[level]
	contextStr := ""
	if context != "" {
		contextStr = fmt.Sprintf("[%s] ", context)
	}
	logLine := fmt.Sprintf("%s %s %s%s", timestamp, levelStr, contextStr, message)

	// Console output
	if globalLogger.consoleOutput {
		color := logLevelColors[level]
		fmt.Printf("%s%s%s\n", color, logLine, colorReset)
	}

	// File output with rotation check
	// First, check if there's a context-specific log file
	if context != "" {
		if contextPath, ok := globalLogger.contextLogFiles[context]; ok {
			file, err := os.OpenFile(contextPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
			if err == nil {
				file.WriteString(logLine + "\n")
				file.Close()
			}
			return // Don't write to default log file if context file is used
		}
	}

	if globalLogger.baseLogFilePath != "" {
		// Check if we need to rotate to a new file
		newRotatedPath := getRotatedFilePath()
		if newRotatedPath != globalLogger.currentRotatedPath {
			globalLogger.currentRotatedPath = newRotatedPath
			globalLogger.logFilePath = newRotatedPath

			// Write header to new rotated file
			if _, err := os.Stat(globalLogger.logFilePath); os.IsNotExist(err) {
				timestamp := time.Now().Format("2006-01-02 15:04:05")
				header := fmt.Sprintf("=== Log started at %s ===\n", timestamp)
				os.WriteFile(globalLogger.logFilePath, []byte(header), 0644)
			}
		}

		if globalLogger.logFilePath != "" {
			file, err := os.OpenFile(globalLogger.logFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
			if err == nil {
				file.WriteString(logLine + "\n")
				file.Close()
			}
		}
	}
}

// EnableFileLogging enables logging to the specified file
func EnableFileLogging(filePath string) {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()

	globalLogger.logFilePath = filePath
	timestamp := time.Now().Format("2006-01-02 15:04:05")
	header := fmt.Sprintf("=== Log started at %s ===\n", timestamp)
	err := os.WriteFile(filePath, []byte(header), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to initialize log file: %v\n", err)
	}
}

// DisableFileLogging disables logging to file
func DisableFileLogging() {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()
	globalLogger.logFilePath = ""
}

// EnableConsoleOutput enables console output
func EnableConsoleOutput() {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()
	globalLogger.consoleOutput = true
}

// DisableConsoleOutput disables console output
func DisableConsoleOutput() {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()
	globalLogger.consoleOutput = false
}

// ConfigureContextLogFiles configures context-specific log files
func ConfigureContextLogFiles(contextLogFiles map[string]string) {
	globalLogger.mutex.Lock()
	defer globalLogger.mutex.Unlock()

	globalLogger.contextLogFiles = make(map[string]string)
	for k, v := range contextLogFiles {
		globalLogger.contextLogFiles[k] = v
	}

	// Initialize each context log file
	for context, path := range contextLogFiles {
		// Create directory if it doesn't exist
		dir := filepath.Dir(path)
		if dir != "" {
			os.MkdirAll(dir, 0755)
		}

		timestamp := time.Now().Format("2006-01-02 15:04:05")
		header := fmt.Sprintf("=== Log started at %s [%s] ===\n", timestamp, context)
		err := os.WriteFile(path, []byte(header), 0644)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Failed to initialize log file for context %s: %v\n", context, err)
		}
	}
}

File: rust/Assembler/src/template_common/logger.rs
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::sync::Mutex;
use std::collections::HashMap;
use chrono::Local;
use std::path::Path;

/// Simple custom logger for consistent logging across all language implementations.
/// Provides file-based logging with different log levels.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum LogLevel {
    DEBUG = 0,
    INFO = 1,
    WARN = 2,
    ERROR = 3,
    NONE = 4,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LogRotation {
    NONE = 0,
    HOURLY = 1,
    DAILY = 2,
}

struct LoggerConfig {
    current_log_level: LogLevel,
    log_file_path: Option<String>,
    console_output: bool,
    log_rotation: LogRotation,
    base_log_file_path: Option<String>,
    current_rotated_path: Option<String>,
    context_log_files: HashMap<String, String>,
}

lazy_static::lazy_static! {
    static ref LOGGER_CONFIG: Mutex<LoggerConfig> = Mutex::new(LoggerConfig {
        current_log_level: LogLevel::NONE,
        log_file_path: None,
        console_output: true,
        log_rotation: LogRotation::NONE,
        base_log_file_path: None,
        current_rotated_path: None,
        context_log_files: HashMap::new(),
    });
}

pub struct Logger;

impl Logger {
    /// Configure the logger
    pub fn configure(level: LogLevel, log_file_path: Option<String>, console_output: bool, rotation: LogRotation) {
        let mut config = LOGGER_CONFIG.lock().unwrap();
        config.current_log_level = level;
        config.base_log_file_path = log_file_path.clone();
        config.console_output = console_output;
        config.log_rotation = rotation;

        // Generate the initial rotated path
        config.current_rotated_path = Self::get_rotated_file_path_internal(&config);
        config.log_file_path = config.current_rotated_path.clone();

        // Create or clear log file if specified
        if let Some(path) = &config.log_file_path {
            if let Ok(mut file) = File::create(path) {
                let _ = writeln!(file, "=== Log started at {} ===", Local::now().format("%Y-%m-%d %H:%M:%S"));
            } else {
                eprintln!("Failed to initialize log file: {}", path);
            }
        }
    }

    /// Generate the rotated file path based on rotation setting
    fn get_rotated_file_path_internal(config: &LoggerConfig) -> Option<String> {
        let base_path = config.base_log_file_path.as_ref()?;

        if config.log_rotation == LogRotation::NONE {
            return Some(base_path.clone());
        }

        let path = Path::new(base_path);
        let directory = path.parent().map(|p| p.to_string_lossy().to_string()).unwrap_or_default();
        let file_stem = path.file_stem().map(|s| s.to_string_lossy().to_string()).unwrap_or_default();
        let extension = path.extension().map(|e| format!(".{}", e.to_string_lossy())).unwrap_or_default();

        let now = Local::now();
        let suffix = match config.log_rotation {
            LogRotation::HOURLY => now.format("%Y-%m-%d_%H").to_string(),
            LogRotation::DAILY => now.format("%Y-%m-%d").to_string(),
            LogRotation::NONE => String::new(),
        };

        let rotated_file_name = format!("{}_{}{}", file_stem, suffix, extension);

        if directory.is_empty() {
            Some(rotated_file_name)
        } else {
            Some(format!("{}/{}", directory, rotated_file_name))
        }
    }

    /// Configure context-specific log files
    pub fn configure_context_log_files(context_log_files: HashMap<String, String>) {
        let mut config = LOGGER_CONFIG.lock().unwrap();
        config.context_log_files = context_log_files.clone();

        // Initialize each context log file
        for (context, path) in &context_log_files {
            // Create directory if it doesn't exist
            if let Some(parent) = Path::new(path).parent() {
                let _ = std::fs::create_dir_all(parent);
            }

            if let Ok(mut file) = File::create(path) {
                let _ = writeln!(file, "=== Log started at {} [{}] ===",
                    Local::now().format("%Y-%m-%d %H:%M:%S"), context);
            } else {
                eprintln!("Failed to initialize log file for context {}: {}", context, path);
            }
        }
    }

    /// Set the current log level
    pub fn set_log_level(level: LogLevel) {
        let mut config = LOGGER_CONFIG.lock().unwrap();
        config.current_log_level = level;
    }

    /// Get the current log level
    pub fn get_log_level() -> LogLevel {
        let config = LOGGER_CONFIG.lock().unwrap();
        config.current_log_level
    }

    /// Log a DEBUG level message
    pub fn debug(message: &str, context: Option<&str>) {
        Self::log(LogLevel::DEBUG, message, context);
    }

    /// Log an INFO level message
    pub fn info(message: &str, context: Option<&str>) {
        Self::log(LogLevel::INFO, message, context);
    }

    /// Log a WARN level message
    pub fn warn(message: &str, context: Option<&str>) {
        Self::log(LogLevel::WARN, message, context);
    }

    /// Log an ERROR level message
    pub fn error(message: &str, context: Option<&str>) {
        Self::log(LogLevel::ERROR, message, context);
    }

    /// Core logging method
    fn log(level: LogLevel, message: &str, context: Option<&str>) {
        let mut config = LOGGER_CONFIG.lock().unwrap();

        if level < config.current_log_level {
            return;
        }

        let timestamp = Local::now().format("%Y-%m-%d %H:%M:%S%.3f");
        let level_str = format!("{:?}", level);
        let level_str_padded = format!("{:<5}", level_str);
        let context_str = context.map(|c| format!("[{}] ", c)).unwrap_or_default();
        let log_line = format!("{} {} {}{}", timestamp, level_str_padded, context_str, message);

        // Console output
        if config.console_output {
            match level {
                LogLevel::DEBUG => println!("\x1b[90m{}\x1b[0m", log_line),  // Gray
                LogLevel::INFO => println!("{}", log_line),                   // White
                LogLevel::WARN => println!("\x1b[33m{}\x1b[0m", log_line),   // Yellow
                LogLevel::ERROR => println!("\x1b[31m{}\x1b[0m", log_line),  // Red
                LogLevel::NONE => {},
            }
        }

        // File output with rotation check
        // First, check if there's a context-specific log file
        if let Some(ctx) = context {
            if let Some(context_path) = config.context_log_files.get(ctx) {
                if let Ok(mut file) = OpenOptions::new()
                    .create(true)
                    .append(true)
                    .open(context_path)
                {
                    let _ = writeln!(file, "{}", log_line);
                }
            }
        } else if config.base_log_file_path.is_some() {
            // Check if we need to rotate to a new file
            let new_rotated_path = Self::get_rotated_file_path_internal(&config);
            if new_rotated_path != config.current_rotated_path {
                config.current_rotated_path = new_rotated_path.clone();
                config.log_file_path = new_rotated_path.clone();

                // Write header to new rotated file
                if let Some(path) = &config.log_file_path {
                    if !std::path::Path::new(path).exists() {
                        if let Ok(mut file) = File::create(path) {
                            let _ = writeln!(file, "=== Log started at {} ===", Local::now().format("%Y-%m-%d %H:%M:%S"));
                        }
                    }
                }
            }

            if let Some(path) = &config.log_file_path {
                if let Ok(mut file) = OpenOptions::new().create(true).append(true).open(path) {
                    let _ = writeln!(file, "{}", log_line);
                }
            }
        }
    }

    /// Enable logging to file
    pub fn enable_file_logging(file_path: &str) {
        let mut config = LOGGER_CONFIG.lock().unwrap();
        config.log_file_path = Some(file_path.to_string());

        if let Ok(mut file) = File::create(file_path) {
            let _ = writeln!(file, "=== Log started at {} ===", Local::now().format("%Y-%m-%d %H:%M:%S"));
        } else {
            eprintln!("Failed to initialize log file: {}", file_path);
        }
    }

    /// Disable logging to file
    pub fn disable_file_logging() {
        let mut config = LOGGER_CONFIG.lock().unwrap();
        config.log_file_path = None;
    }

    /// Enable console output
    pub fn enable_console_output() {
        let mut config = LOGGER_CONFIG.lock().unwrap();
        config.console_output = true;
    }

    /// Disable console output
    pub fn disable_console_output() {
        let mut config = LOGGER_CONFIG.lock().unwrap();
        config.console_output = false;
    }
}
